### ISSUE TITLE
Packaging

### ISSUE THREAD

**Issue by oleastre** [hanatos/vkdt#39]
As explained on pixls.us, I already did a package of vkdt for fedora, and I'm now trying to create a flatpak one.
I open this issue to centralise a bit all issues I found along the way. Here is a simple list:

- have a reverse DNS name, used in `.desktop` file and or flatpak, ... useful to have unique app id
- have `.desktop` file with probably an icon
- have an appdata file that market the app purpose, where to find doc, report issues, ...

I'm currently trying to get a clang compiler working with LLVM to b able to compile the project.

**Comment by hanatos** [comment]
..if it's any help, gcc should work too (just need to fill it in in the appropriate spot in `bin/config.mk` overwriting whatever was set in `bin/config.mk.defaults`. i do that for some of the rpm packages on obs.

**Comment by podiki** [comment]
I'm not sure if this fits here since maybe this is just for Flatpak packaging, but I have a package for GNU Guix that works for me. I'll get it submitted upstream or make it available on my channel and share the link.

Would it be useful to have a `guix.scm` file that one can use to build vkdt and/or setup a dev environment with the needed dependencies? This would use (the really great) [Guix shell](https://guix.gnu.org/manual/devel/en/html_node/Invoking-guix-shell.html) so one can just run `guix shell` in the cloned repo and be ready to hack or run vkdt. The advantage, besides being very easy, is that this could be used on just about any Linux distro (just need to install Guix).

**Comment by dbaynard** [comment]
> I have a package for GNU Guix that works for me

This would be amazing, for me! I was going to take a look at packaging for `nix`, so I can use vkdt on NixOS — I’d love to build on your efforts.

**Comment by podiki** [comment]
@dbaynard Here is my current WIP, hope it is helpful: https://gitlab.com/podiki/guix-pod/-/blob/main/vkdt.scm

The build is pretty straightforward and although I don't know how to use much of vkdt, it seems to generally work. I'm seeing a few bugs and things I would want to tweak, but I think that is in vkdt rather than packaging.

**Comment by hanatos** [comment]
"I'm seeing a few bugs and things I would want to tweak,"

do let me know about that, i might be able to fix it 😃

**Comment by paperdigits** [comment]
@dbaynard I've started a package for NixOS: https://github.com/paperdigits/nixpkgs/commit/9d9687b554e2696c052fa533e83737f17d0c0e8a

It isn't quite working yet. I've been trying to just build vkdt inside a nix shell, and while it builds successfully, it doesn't launch:

```
nix-shell -p vulkan-headers vulkan-tools vulkan-loader glfw libjpeg pkg-config rsync cmake freetype exiv2 ffmpeg clang pugixml llvm llvmPackages.openmp glslang vulkan-tools-lunarg vulkan-validation-layers
cd bin
make debug -j12
```

Then run it:

```
[nix-shell:~/Desktop/vkdt/bin]$ ./vkdt -d all ~/Desktop/Photographs/Raw_files/2021-05-21/2021-05-21_15-38-55_DSC1790.nef
[gui] glfwGetVersionString() : 3.3.7 X11 GLX EGL OSMesa clock_gettime evdev shared
[gui] monitor [0] eDP-1 at 0 0
[gui] vk extension required by GLFW:
[gui]   VK_KHR_surface
[gui]   VK_KHR_xcb_surface
[qvk] error VK_ERROR_LAYER_NOT_PRESENT executing vkCreateInstance(&inst_create_info, NULL, &qvk.instance)!
[ERR] init vulkan failed
[ERR] failed to init gui/swapchain
```

**Comment by hanatos** [comment]
hi mica and thanks for the effort!

"layer not present" often indicates that it doesn't find the validation
layers required for debug mode. since you did install it i'm not sure
what's going on. since `VK_LAYER_KHRONOS_validation` is in fact the only
(optional, only in debug mode) layer that is used, maybe there's an issue
with paths/the setup? does the plain `make -j12` version without debug work?

**Comment by podiki** [comment]
It is probably a question for Nix packagers that know how to handle accessing Vulkan stuff. Maybe look at some packages there that also use Vulkan to see what they do (sorry don't know much about Nix, and not sure if building in a nix shell rather than as a package definition is easier or harder)

**Comment by dbaynard** [comment]
From a cursory search, it seems similar issues with layers have been fixed by setting environment variables explicitly — the implication being that the nix-shell doesn't get this right.

https://discourse.nixos.org/t/setting-up-vulkan-for-development/11715 and https://www.reddit.com/r/NixOS/comments/npxlql/unstable_channel_vulkan_validation_layer_problems/

This looks great, though, and you definitely beat me to it!

**Comment by paperdigits** [comment]
@dbaynard you're right, it is that an environment variable was not set correctly.

Using the following `shell.nix` file, I'm able to build both the debug and non-debug version of `vkdt`. Strangely, `rawspeed` doesn't seem to be able to decode files from my D850, but I fed it an older `.cr2` file and it opened. 

Anyway, here is the current `shell.nix`:

```
{ pkgs ? import <nixpkgs> { }, lib ? pkgs.lib }:
pkgs.mkShell rec {
  name = "vkdt-dev-shell";
  buildInputs = with pkgs; [
    vulkan-headers
    vulkan-tools
    vulkan-loader
    glfw
    libjpeg
    pkg-config
    rsync
    cmake
    freetype
    exiv2
    ffmpeg
    clang
    pugixml
    llvm
    llvmPackages.openmp
    glslang
    vulkan-tools-lunarg
    vulkan-validation-layers
  ];
  #LD_LIBRARY_PATH = "${lib.makeLibraryPath buildInputs}";
  VK_LAYER_PATH = "${pkgs.vulkan-validation-layers}/share/vulkan/explicit_layer.d";
  #XDG_DATA_DIRS = builtins.getEnv "XDG_DATA_DIRS";
  #XDG_RUNTIME_DIR = builtins.getEnv "XDG_RUNTIME_DIR";
}
```

**Comment by paperdigits** [comment]
@dbaynard I now have vkdt building and launch from nixpkgs: https://github.com/paperdigits/nixpkgs/commit/8afc5e09e3d4ac315327f8233830962adc2f2c94

I need to do a bit more to get it to `$out/bin` instead of `$out/usr/bin` but that'll be trivial.

Thanks @hanatos !

I'll probably wait just a bit to actually submit the PR to nixpkgs, as @hanatos mentioned maybe a 0.5 release might be soon...

**Comment by dbaynard** [comment]
Ah @paperdigits this is great! What platforms have you tested — just `x86_64-linux`?

**Comment by hanatos** [comment]
heya,

fwiw the makefile obeys the std semantics of `DESTDIR` and `prefix`. the directory listing you're posting looks like `DESTDIR=result` and `prefix=/usr` but you can of course change them as you please. the only thing that is essential for vkdt to function correctly is that `bin/` contains symlinks to the real binary in what's `result/lib/` here (i.e. the place where `data/` and `modules/` can be found as subfolders).

also i just tagged a release :)

**Comment by paperdigits** [comment]
Thanks @hanatos !

@dbaynard I think this is the one, if you don't mind taking a look https://github.com/paperdigits/nixpkgs/commit/aee98502868e77fe46911d57a6433fc3666e580c

**Comment by hanatos** [comment]
.. i think this is done, right? if not please reopen or start another.

**Comment by dbaynard** [comment]
Yes, probably more of a discussion thread than a task for maintainers, at this point. I missed the comment from @paperdigits — I'll have to take another look, as asked.

**Comment by dbaynard** [comment]
@paperdigits I can export but I don't see previews, though the only error I see is that `/usr/bin/exiftool` can't be found. For the latter I think we'll need to patch `src/gui/render_lighttable.cc` (and might want to make the path to exiftool customizable). For the former, I'm not sure what's going on.

```
[gui] glfwGetVersionString() : 3.3.8 X11 GLX EGL OSMesa clock_gettime evdev shared
[gui] monitor [0] eDP-1 at 0 0
[gui] vk extension required by GLFW:
[gui]   VK_KHR_surface
[gui]   VK_KHR_xcb_surface
[gui] found joystick FRMW0001:00 32AC:0006 Consumer Control
[gui] disabling joystick due to explicit config request. enable by
[gui] setting 'intgui/disable_joystick:0' in ~/.config/vkdt/config.rc
[gui] no display profile file display.eDP-1, using sRGB!
[gui] no display profile file display.eDP-1, using sRGB!
```

**Comment by paperdigits** [comment]
VKDT is in nixpkgs now, not sure if it has hit unstable yet, but it was merged onto master a few weeks ago.

**Comment by dbaynard** [comment]
https://nixpk.gs/pr-tracker.html?pr=214585 shows 0.5.4 PR has been merged, too — thanks!

**Comment by paperdigits** [comment]
Hmmm I didn't see that error on my machine. I'll have to look again.

**Comment by hanatos** [comment]
these outputs are just informative messages and don't mean errors. the thumbnail render issue was present in the initial release but should be fixed in 0.5.4. let me know if it is in the latest version and i'll have to look again. exiftool is just used for metadata display in the expander in lighttable mode (kinda optional). the path is user configurable (along with the whole command) in `~/.config/vkdt/config.rc`.
